eth/fetcher/fetcher.go

```diff
@@ -682,25 +695,17 @@ func (f *Fetcher) insert(peer string, block *types.Block) {
 	// Run the import on a new thread
 	glog.V(logger.Debug).Infof("Peer %s: importing block #%d [%s]", peer, block.NumberU64(), hash.Hex())
 	go func() {
-
-		haveParent := true
-
-		defer func() {
-			if haveParent {
-				f.done <- hash
-			}
-		}()
+		defer func() { f.done <- hash }()
 
 		// If the parent's unknown, abort insertion, and don't forget the hash and block;
 		// use queue gap fill to get unknown parent.
 		parent := f.getBlock(block.ParentHash())
 		if parent == nil {
 			glog.V(logger.Debug).Infof("Peer %s: parent [%s] of block #%d [%s] unknown", peer, block.ParentHash().Hex(), block.NumberU64(), hash.Hex())
-			haveParent = false
 			return
 		}
 		// Quickly validate the header and propagate the block if it passes
-		switch err := f.validateBlock(block, parent); err {
+		switch err := f.verifyHeader(block.Header()); err {
```

Comment:
Just a quick note that in case you were wondering about the where the hell the parent went -- the validator fn passed in from the handler sets this per the blockchain validator, moving the responsibility for parent-aware validation logic back to the chain validator itself.

```go
	validator := func(header *types.Header) error {
		return manager.blockchain.Validator().ValidateHeader(header, manager.blockchain.GetHeader(header.ParentHash), true)
	}
```
> eth/handler.go#168